### DISCUSSION TITLE
Global Event Bus in Quasar v2

### DISCUSSION THREAD

**Discussion by h3** [quasarframework/quasar#9220] [General - Components / Directives / etc] [answered]
In Quasar v1 there was the [Global Event Bus](https://quasar.dev/options/global-event-bus).

I'm currently in the process of upgrading an application which makes use of it and noticed it is not documented in the [new documentation](https://next.quasar.dev/options/global-event-bus) nor mentioned in the [upgrade guide](https://next.quasar.dev/start/upgrade-guide).

It is however still referenced in the [Handling Vue Events section](https://next.quasar.dev/start/how-to-use-vue#handling-vue-events) .. 

Is it deprecated? If so what is the proper upgrade path to handle legacy code that used it?

**Comment by metalsadman** [comment]
Yeah it should be removed, for migration guide seek out vue docs about this here https://v3.vuejs.org/guide/migration/events-api.html#events-api.

**Reply by h3** [reply]
Thanks!

**Reply by h3** [reply]
For those looking for a near drop-in substitution just create a boot file:

```
$ yarn add tiny-emitter
```

### src/boot/global-event-bus.js

```javascript
import emitter from 'tiny-emitter/instance'

export default ({ app }) => {
  app.config.globalProperties.$global = {
    $on: (...args) => emitter.on(...args),
    $once: (...args) => emitter.once(...args),
    $off: (...args) => emitter.off(...args),
    $emit: (...args) => emitter.emit(...args)
  }
}
```

Add `'global-event-bus'` to the `boot` section in `quasar.conf.js`

**Note**: I used `$global` because extending `$root` does not seem to work.

Bonus vim substitution one-liner: `%s/$root\.$\(on\|off\|once\|emit\)/$global\.$\1/g`

**Reply by ptcampos** [reply]
Thank you man! Saved my day. :+1:

**Reply by joseseie** [reply]
Yes, it works fine `this.$global`;

**Reply by Junzlay** [reply]
how to use this in composition api?

**Reply by lawrence615** [reply]
> how to use this in composition api?

@Junzlay did you figure out how to use it?

**Reply by ghost** [reply]
@Junzlay @lawrence615 I was able to solve it using injection: https://stackoverflow.com/questions/63100658/add-global-variable-in-vue-js-3

It's actually available in the templates but to use them in the setup script you'll need to inject them. There are a few ways to do this but what I did was change `/src/boot/global-event-bus.js` to this:

```js
import emitter from 'tiny-emitter/instance'

export default ({ app }) => {
  app.provide('$global', {
    $on: (...args) => emitter.on(...args),
    $once: (...args) => emitter.once(...args),
    $off: (...args) => emitter.off(...args),
    $emit: (...args) => emitter.emit(...args)
  })
}
```

Then, in my component:

```js
<script setup>
import {inject} from 'vue'
const $global = inject('$global')
console.log($global)
</script>
```

**Reply by lawrence615** [reply]
@EverestWonder Thanks.

**Comment by hawkeye64** [comment]
https://github.com/quasarframework/quasar/pull/9225

**Comment by rstoenescu** [comment, marked as answer]
A Quasar native solution (through EventBus util) will be available in v2.8.4. Sneak [docs preview](https://github.com/quasarframework/quasar/blob/dev/docs/src/pages/quasar-utils/event-bus-util.md) (until releasing it). For those who read this after v2.8.4 has been released: [docs page](https://quasar.dev/quasar-utils/event-bus-util). It also has examples with Options API and Composition API.
